### PR TITLE
tools: add fpgainfo for Darby Creek

### PR DESCRIPTION
**Pull request by anandaravuri**
- Added libboard_dc 
- Added test cases

### CODE REVIEW

File: testing/board/test_board_dc.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+	char buf[10] = {0};
+	char buf_ver[10] = {0};
+
+	EXPECT_EQ(parse_fw_ver(buf,buf_ver,strlen(buf)), FPGA_OK);
+
+	EXPECT_EQ(parse_fw_ver(NULL, buf_ver, strlen(buf)), FPGA_INVALID_PARAM);
+	EXPECT_EQ(parse_fw_ver(buf, NULL, strlen(buf)), FPGA_INVALID_PARAM);
+
+	EXPECT_EQ(parse_fw_ver(buf, buf_ver, 0x100000), FPGA_OK);
```
Comment (jelonanX) [inline testing/board/test_board_dc.cpp]: would it be helpful to initialize the buffer to a known value and add another check after the FPGA_OK cases that compares what was parsed to what was fed in?

File: testing/board/test_board_dc.cpp [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+* @details    Validates read mac info  <br>
+*/
+TEST_P(board_dc_c_p, board_dc_6) {
+
+	struct ether_addr mac_addr;
+
+	EXPECT_EQ(read_mac_info(tokens_[0], 0, &mac_addr), FPGA_OK);
+	EXPECT_EQ(read_mac_info(tokens_[0], 0, NULL), FPGA_INVALID_PARAM);
+
+	EXPECT_EQ(read_mac_info(tokens_[0], 200, &mac_addr), FPGA_INVALID_PARAM);
```
Comment (jelonanX) [inline testing/board/test_board_dc.cpp]: why does this return FPGA_INVALID_PARAM?

File: tools/base/libboard/board_dc/board_dc.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+*/
+fpga_result read_bmcfw_version(fpga_token token, char *bmcfw_var, size_t len);
+
+/**
+* Parse bmc/max10 version.
+*
+* @param[in] buf             pointer to firmware version
+* @param[inout] fw_ver       pointer to char firmware string
+* @param[in] len             length of char fw var string
+* @returns FPGA_OK on success. FPGA_EXCEPTION if FW version is invlaid.
```
Comment (jelonanX) [inline tools/base/libboard/board_dc/board_dc.h]: s/invlaid/invalid

File: tools/base/libboard/board_dc/board_dc.c [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+	fpga_object bmcfw_object;
+
+	if (bmcfw_ver == NULL) {
+		FPGA_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_BMCFW_VER, &bmcfw_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
```
Comment (jelonanX) [inline tools/base/libboard/board_dc/board_dc.c]: should this be an OPAE_ERR?

File: tools/base/libboard/board_dc/board_dc.c [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+	fpga_object max10fw_object;
+
+	if (max10fw_ver == NULL) {
+		FPGA_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_MAX10_VER, &max10fw_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
```
Comment (jelonanX) [inline tools/base/libboard/board_dc/board_dc.c]: should this be an OPAE_ERR?

File: testing/board/test_board_dc.cpp [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+	void *buf, size_t count) {
+	fpga_result res = FPGA_OK;
+	char sysfspath[FPGA_VAR_BUF_LEN];
+	int fd = 0;
+
+	snprintf_s_ss(sysfspath, sizeof(sysfspath), "%s/%s", "/sys/class/fpga/intel-fpga-dev.0/intel-fpga-fme.0", file);
+
+	glob_t pglob;
+	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
+	if ((gres) || (1 != pglob.gl_pathc)) {
```
Comment (tswhison) [inline testing/board/test_board_dc.cpp]: Shouldn't this check pglob.gl_pathv for NULL?

File: testing/board/test_board_dc.cpp [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+fpga_result board_dc_c_p::delete_sysfs_file(const char *file) {
+	fpga_result res = FPGA_OK;
+	char sysfspath[FPGA_VAR_BUF_LEN];
+	int status = 0;
+
+	snprintf_s_ss(sysfspath, sizeof(sysfspath), "%s/%s", "/sys/class/fpga/intel-fpga-dev.0/intel-fpga-fme.0", file);
+
+	glob_t pglob;
+	int gres = glob(sysfspath, GLOB_NOSORT, NULL, &pglob);
+	if ((gres) || (1 != pglob.gl_pathc)) {
```
Comment (tswhison) [inline testing/board/test_board_dc.cpp]: Check pglob.pathv?

File: testing/board/test_board_dc.cpp [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+* @test       board_dc_5
+* @brief      Tests: parse_fw_ver
+* @details    Validates parse fw version  <br>
+*/
+TEST_P(board_dc_c_p, board_dc_5) {
+
+	char buf[10] = {0};
+	char buf_ver[10] = {0};
+
+	EXPECT_EQ(parse_fw_ver(buf,buf_ver,strlen(buf)), FPGA_OK);
```
Comment (tswhison) [inline testing/board/test_board_dc.cpp]: If the 3rd parameter is the max number of characters to copy, then strlen(buf) will produce the wrong answer.  You'll want sizeof(buf_ver).

File: tools/base/libboard/board_dc/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,66 @@
+## Copyright(c) 2019, Intel Corporation
```
Comment (tswhison) [inline tools/base/libboard/board_dc/CMakeLists.txt]: Need to update opae.spec.in, too?

File: tools/base/libboard/board_dc/board_dc.c [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+	res = fpgaObjectGetSize(bmcfw_object, &size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object size ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	// Return error if object size bigger then buffer size
+	if (size > FPGA_VAR_BUF_LEN) {
+		FPGA_ERR("object size bigger then buffer size");
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: FPGA_ERR or OPAE_ERR ?

File: tools/base/libboard/board_dc/board_dc.c [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	res = parse_fw_ver(buf, bmcfw_ver, len);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to parse version ");
+		resval = res;
+		goto out_destroy;
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: This goto isn't necessary.

File: tools/base/libboard/board_dc/board_dc.c [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+fpga_result parse_fw_ver(char *buf, char *fw_ver, size_t len)
+{
+	uint32_t var               = 0;
+	fpga_result res            = FPGA_OK;
+	int retval                 = 0;
+
+	if (buf == NULL ||
+		fw_ver == NULL) {
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: alignment is off.

File: tools/base/libboard/board_dc/board_dc.c [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+	/* BMC FW version format reading
+	NIOS II Firmware Build 0x0 32 RW[23:0] 24 hFFFFFF Build version of NIOS II Firmware
+	NIOS FW is up e.g. 1.0.1 for first release
+	[31:24] 8hFF Firmware Support Revision - ASCII code
+	0xFF is the default value without NIOS FW, will be changed after NIOS FW is up
+	*/
+
+	errno = 0;
+	var = strtoul(buf, NULL, 16);
+	if (var == 0  &&
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: Do this to check the result of strtoul():
```c
char *endptr = NULL;
var = strtoul(buf, &endptr, 16);
if (endptr != buf + strnlen(buff, 100)) {
  // strtoul() failed.
}
```

File: tools/base/libboard/board_dc/board_dc.c [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	res = parse_fw_ver(buf, max10fw_ver, len);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to parse version ");
+		resval = res;
+		goto out_destroy;
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: This goto isn't needed.

File: tools/base/libboard/board_dc/board_dc.c [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+	fpga_result res                      = FPGA_OK;
+	char bmc_ver[FPGA_VAR_BUF_LEN]       = { 0 };
+	char max10_ver[FPGA_VAR_BUF_LEN]     = { 0 };
+	char mac_str[18] = { 0 };
+	struct ether_addr MAC ;
+
+
+	res = read_bmcfw_version(token, bmc_ver, FPGA_VAR_BUF_LEN);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to read bmc version");
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: OPAE_ERR?

File: tools/base/libboard/board_dc/board_dc.c [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+	}
+
+	res = read_max10fw_version(token, max10_ver, FPGA_VAR_BUF_LEN);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to read max10 version");
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: OPAE_ERR?

File: tools/base/libboard/board_dc/board_dc.c [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+	}
+
+	printf("Board Management Controller, MAX10 NIOS FW version: %s \n", bmc_ver);
+	printf("Board Management Controller, MAX10 Build version: %s \n", max10_ver);
+
+	memset_s((void*)&MAC, sizeof(MAC), 0);
+
+	res = read_mac_info(token, 0, &MAC);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to read mac address");
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.c]: OPAE_ERR?

File: tools/base/libboard/board_dc/board_dc.h [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef __FPGA_BOARD_VC_H__
+#define __FPGA_BOARD_VC_H__
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: s/__FPGA_BOARD_VC_H__/__FPGA_BOARD_DC_H__/

File: tools/base/libboard/board_dc/board_dc.h [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+* FPGA_INVALID_PARAM if invalid parameters were provide
+*
+*/
+fpga_result print_board_info(fpga_token token);
+
+/**
+* Get Max10 firmware version.
+*
+* @param[in] token           fpga_token object for device (FPGA_DEVICE type)
+* @param[inout] max10fw_var  pointer to char pcb_info string
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: What is "pcb_info string" referring to?

File: tools/base/libboard/board_dc/board_dc.h [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+* Get BMC/NIOS firmware version.
+*
+* @param[in] token           fpga_token object for device (FPGA_DEVICE type)
+* @param[inout] bmcfw_var    pointer to char bmcfw_var string
+*                            user allocates memory and free input string
+* @param[in] len             length of char bmcfw_var string
+* @returns FPGA_OK on success. FPGA_NOT_FOUND if NIOS sysfs not found.
+* FPGA_INVALID_PARAM if invalid parameters were provide
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: s/provide/provided/

File: tools/base/libboard/board_dc/board_dc.h [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+* Get mac address.
+*
+* @param[in] token           fpga_token object for device (FPGA_DEVICE type)
+* @param[in] afu_channel_num afu channel number
+* @param[inout] mac_addr  pointer to struct ether_addr
+
+* @returns FPGA_OK on success. FPGA_NOT_FOUND if invalid MAC address.
+* FPGA_INVALID_PARAM if invalid parameters were provide
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: s/provide/provided/

File: tools/base/libboard/board_dc/board_dc.h [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+* Get Max10 firmware version.
+*
+* @param[in] token           fpga_token object for device (FPGA_DEVICE type)
+* @param[inout] max10fw_var  pointer to char pcb_info string
+*                            user allocates memory and free input string
+* @param[in] len             length of char max10fw_var string
+* @returns FPGA_OK on success. FPGA_NOT_FOUND if MAX10  sysfs not found.
+* FPGA_INVALID_PARAM if invalid parameters were provide
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: s/provide/provided/

File: tools/base/libboard/board_dc/board_dc.h [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+*
+*/
+fpga_result read_mac_info(fpga_token token, uint32_t afu_channel_num,
+				struct ether_addr *mac_addr);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+#endif /* __FPGA_BOARD_VC_H__ */
```
Comment (tswhison) [inline tools/base/libboard/board_dc/board_dc.h]: s/__FPGA_BOARD_VC_H__/__FPGA_BOARD_DC_H__/

Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: Fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 22 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 21 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 20 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 19 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 14 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 15 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 17 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 18 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 16 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 12 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 11 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 10 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/CMakeLists.txt (hunk 9 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/CMakeLists.txt]: fixed

File: testing/board/test_board_dc.cpp (hunk 8 above)
Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: fixed

File: testing/board/test_board_dc.cpp (hunk 7 above)
Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: fixed

File: testing/board/test_board_dc.cpp (hunk 6 above)
Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 5 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.c (hunk 4 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.c]: fixed

File: tools/base/libboard/board_dc/board_dc.h (hunk 3 above)
Comment (anandaravuri) [reply tools/base/libboard/board_dc/board_dc.h]: fixed

File: testing/board/test_board_dc.cpp (hunk 2 above)
Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: fixed

Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: Invalid AFU channel

File: testing/board/test_board_dc.cpp (hunk 1 above)
Comment (anandaravuri) [reply testing/board/test_board_dc.cpp]: fixed

File: scripts/cover.sh [hunk 24]
```diff
@@ -35,5 +35,5 @@ lcov -c -i -d . -o coverage.base 2> /dev/null
 LD_LIBRARY_PATH=${PWD}/lib \
 CTEST_OUTPUT_ON_FAILURE=1 \
 OPAE_EXPLICIT_INITIALIZE=1 \
-ctest -j3 --timeout 180
+ctest -j3 --timeout 300
```
[the remark this answers is not in the document]
Comment (anandaravuri) [reply scripts/cover.sh]: Added new mock sysfs for DC card.

**Comment by tswhison** [comment]
ping

**Pull request closed without merging**